### PR TITLE
Client side saving

### PR DESCRIPTION
**Pull request by NancokPS2**
For use with the options menu atm.   
Reserved for client-only data, like settings or any data that the server doesn't need to concern itself with.

### CODE REVIEW

File: scripts/classes/LocalSaving.gd [hunk 1]
```diff
@@ -0,0 +1,28 @@
+extends Object
+class_name LocalSaveSystem
+
+const DEFAULT_PATH:String = "user://SaveFile.ini"
```
Comment (jonathaneeckhout) [inline scripts/classes/LocalSaving.gd:4]: Initially I would have said that I think this file should be placed in the "data" directory, the same as the json database does. But I am not convinced anymore that the data directory is the correct place. Maybe we should create a "user" directory to store all user files which should not be uploaded to git and place a .gitignore file in that directory with the following content:

``` 
# Ignore everything in this directory
*
# Except this file
!.gitignore
```

File: scripts/classes/LocalSaving.gd [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+const Sections:Dictionary = {SETTINGS = "SETTINGS"}
+
+static var savedData:=ConfigFile.new()
```
Comment (jonathaneeckhout) [inline scripts/classes/LocalSaving.gd:8]: Add a space before ConfigFile

File: scripts/classes/LocalSaving.gd (hunk 1 above)
Comment (NancokPS2) [reply scripts/classes/LocalSaving.gd:4]: Currently it is saved to the designated AppData folder that Godot handles, do you want it in the same folder as the game?

**Comment by jonathaneeckhout** [comment]
@NancokPS2 could you rebase on the latest main so I can retest with the menu?

**Comment by NancokPS2** [comment]
Rebased.  
Currently there is nothing in place to use this (since the options didn't exist before this addition)  
To test it, you can run the following script anywhere (this class is fully static):  
    
```
##Try to get an unset value
print( LocalSavingSystem.get_data("SETTINGS", "thisThing", "failsafe") ) ## Prints "failsafe", the 3rd parameter is optional  
  
##Set a value (this can now be retrieved with get)
LocalSavingSystem.set_data( "SETTINGS", "thisThing", "it exists" )    
  
## Save the file (has a default, always accessible path if none is given)
LocalSavingSystem.save_file()    
  
## Manually wipe the stored data (you could also just restart the game to do the same)  
## Yes, the syntax is weird, Godot be like that
LocalSavingSystem.new().savedData = null  
    
## Load what was saved, also uses the same default path
LocalSavingSystem.load_file()  
    
##Try to get the data
print( LocalSavingSystem.get_data("SETTINGS", "thisThing", "failsafe") ) ## Prints "it exists"  
```

**Comment by jonathaneeckhout** [comment]
There is a typo in the code snipped: "LocalSavingSystem" should be "LocalSaveSystem" to match the code. Anway tested it an it seems to work. 

Where is this file actually stored on the filesystem? Which file should I delete to get in a clean state?

Will merge.

**Comment by NancokPS2** [comment]
> There is a typo in the code snipped: "LocalSavingSystem" should be "LocalSaveSystem" to match the code. Anway tested it an it seems to work.
> 
> Where is this file actually stored on the filesystem? Which file should I delete to get in a clean state?
> 
> Will merge.

The directory is OS dependent and always accessible, it is not inside the game folder since not all OSs allow modifying the game directory (and it is a bit hacky since it points to the project folder when using the editor). 
You can find it in the editor with Project>Open User Data Folder  
When i implement its usage i will add an option to reset settings and also include a default settings file in the game's directory to use as a fail safe AND as a reset point.

**Pull request merged**